File: sdk/python/prompt_firewall.py
"""
Prompt Firewall SDK for Python

A lightweight SDK for interacting with the Prompt Firewall API.
"""

import requests
from typing import Optional, Dict, List, Any
from dataclasses import dataclass


@dataclass
class FirewallResult:
    """Result from firewall evaluation"""
    id: str
    decision: str  # 'allow', 'block', 'modify', 'sanitize'
    prompt_modified: str
    response: Optional[str]
    blocked: bool
    risks: List[Dict[str, Any]]
    explain: List[Dict[str, Any]]


class PromptFirewall:
    """Client for Prompt Firewall API"""
    
    def __init__(self, base_url: str = "http://localhost:3000", api_key: Optional[str] = None):
        """
        Initialize the Prompt Firewall client.
        
        Args:
            base_url: Base URL of the firewall API
            api_key: Optional API key for authenticated requests
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.headers = {
            'Content-Type': 'application/json',
        }
        if api_key:
            self.headers['Authorization'] = f'Bearer {api_key}'
    
    def query(self, prompt: str) -> FirewallResult:
        """
        Evaluate a prompt through the firewall.
        
        Args:
            prompt: The user's prompt to evaluate
            
        Returns:
            FirewallResult with decision, risks, and explanations
            
        Example:
            >>> firewall = PromptFirewall()
            >>> result = firewall.query("My email is user@example.com")
            >>> print(result.decision)  # 'sanitize'
            >>> print(result.risks)  # [{'type': 'pii_email', ...}]
        """
        response = requests.post(
            f"{self.base_url}/api/v1/query",
            json={"prompt": prompt},
            headers=self.headers
        )
        response.raise_for_status()
        data = response.json()
        
        if not data.get('success'):
            raise Exception(f"API error: {data.get('error')}")
        
        result_data = data['data']
        return FirewallResult(
            id=result_data['id'],
            decision=result_data['decision'],
            prompt_modified=result_data['promptModified'],
            response=result_data.get('response'),
            blocked=result_data['blocked'],
            risks=result_data.get('risks', []),
            explain=result_data.get('explain', [])
        )
    
    def get_policies(self) -> List[Dict[str, Any]]:
        """Get all firewall policies (requires authentication)"""
        response = requests.get(
            f"{self.base_url}/api/v1/policy",
            headers=self.headers
        )
        response.raise_for_status()
        data = response.json()
        return data.get('data', [])
    
    def get_logs(
        self,
        has_pii: Optional[bool] = None,
        has_injection: Optional[bool] = None,
        blocked: Optional[bool] = None,
        limit: int = 100,
        offset: int = 0
    ) -> Dict[str, Any]:
        """Get firewall logs (requires authentication)"""
        params = {'limit': limit, 'offset': offset}
        if has_pii is not None:
            params['hasPII'] = str(has_pii).lower()
        if has_injection is not None:
            params['hasInjection'] = str(has_injection).lower()
        if blocked is not None:
            params['blocked'] = str(blocked).lower()
        
        response = requests.get(
            f"{self.base_url}/api/v1/logs",
            params=params,
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()


# Example usage
if __name__ == "__main__":
    # Initialize client
    firewall = PromptFirewall(base_url="http://localhost:3000")
    
    # Evaluate a prompt
    result = firewall.query("My email is sarav@example.com")
    print(f"Decision: {result.decision}")
    print(f"Modified prompt: {result.prompt_modified}")
    print(f"Risks detected: {len(result.risks)}")

File: sdk/javascript/prompt-firewall.js
/**
 * Prompt Firewall SDK for JavaScript/Node.js
 * 
 * A lightweight SDK for interacting with the Prompt Firewall API.
 */

class PromptFirewall {
  /**
   * Initialize the Prompt Firewall client
   * @param {string} baseUrl - Base URL of the firewall API
   * @param {string} apiKey - Optional API key for authenticated requests
   */
  constructor(baseUrl = 'http://localhost:3000', apiKey = null) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
    this.headers = {
      'Content-Type': 'application/json',
    };
    if (apiKey) {
      this.headers['Authorization'] = `Bearer ${apiKey}`;
    }
  }

  /**
   * Evaluate a prompt through the firewall
   * @param {string} prompt - The user's prompt to evaluate
   * @returns {Promise<Object>} FirewallResult with decision, risks, and explanations
   * 
   * @example
   * const firewall = new PromptFirewall();
   * const result = await firewall.query("My email is user@example.com");
   * console.log(result.decision); // 'sanitize'
   * console.log(result.risks); // [{type: 'pii_email', ...}]
   */
  async query(prompt) {
    const response = await fetch(`${this.baseUrl}/api/v1/query`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({ prompt }),
    });

    if (!response.ok) {
      throw new Error(`API error: ${response.statusText}`);
    }

    const data = await response.json();
    if (!data.success) {
      throw new Error(`API error: ${data.error}`);
    }

    return data.data;
  }

  /**
   * Get all firewall policies (requires authentication)
   * @returns {Promise<Array>} List of policies
   */
  async getPolicies() {
    const response = await fetch(`${this.baseUrl}/api/v1/policy`, {
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`API error: ${response.statusText}`);
    }

    const data = await response.json();
    return data.data || [];
  }

  /**
   * Get firewall logs (requires authentication)
   * @param {Object} filters - Optional filters
   * @param {boolean} filters.hasPII - Filter by PII presence
   * @param {boolean} filters.hasInjection - Filter by injection presence
   * @param {boolean} filters.blocked - Filter by blocked status
   * @param {number} filters.limit - Number of results (default: 100)
   * @param {number} filters.offset - Pagination offset (default: 0)
   * @returns {Promise<Object>} Logs with pagination info
   */
  async getLogs(filters = {}) {
    const params = new URLSearchParams();
    if (filters.hasPII !== undefined) params.append('hasPII', filters.hasPII);
    if (filters.hasInjection !== undefined) params.append('hasInjection', filters.hasInjection);
    if (filters.blocked !== undefined) params.append('blocked', filters.blocked);
    params.append('limit', filters.limit || 100);
    params.append('offset', filters.offset || 0);

    const response = await fetch(`${this.baseUrl}/api/v1/logs?${params}`, {
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`API error: ${response.statusText}`);
    }

    return response.json();
  }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
  module.exports = PromptFirewall;
}

// Example usage
if (typeof window !== 'undefined') {
  window.PromptFirewall = PromptFirewall;
}

